### DISCUSSION TITLE
Operational models refactoring ideas

### DISCUSSION THREAD

**Discussion by matthewcornell** [reichlab/operational-models#42] [Ideas]
# Intro

This is a starting point for discussing refactoring the code in https://github.com/reichlab/iddata/ , https://github.com/reichlab/idmodels , and this repo ( https://github.com/reichlab/operational-models/ ). The code base developed organically and is now exhibiting some [smells](https://en.wikipedia.org/wiki/Code_smell).

We want to achieve these goals:
- make it easier to add new models, features, and data sources
- make the code cleaner and easier to maintain
- remove code "smells"

Some possible lower level goals:
- centralize/improve error conditions (model or source validation)
- remove duplicated code

# Previous discussions

- [this section](https://docs.google.com/document/d/1W3ixyBB6n2pc9xrnYUbK8FcVrm0Ujp8ZJrZXdShhGSA/edit?tab=t.0#bookmark=id.htuzw17l1fn3) of our dev team agenda
- TBC

# Previous claude analysis
- see [prompt.md](https://github.com/user-attachments/files/25189031/prompt.md) and [analysis.md](https://github.com/user-attachments/files/25189035/analysis.md)

**Comment by matthewcornell** [comment]
Rough notes from today's code review w/ @lshandross and @trobacker .

# NEXT: iddata
- features: probably should not include any feature engineering - should go in idmodels
- sources: kwargs smell. ex: flusurvnet_kwargs=None, nhsn_kwargs=None, ilinet_kwargs=None, nssp_kwargs
- transformations: similar concerns as features. be careful for nssp data source (currently using default one for nhsn)


# operational-models/
## run_config = SimpleNamespace(...)
- two sections - hub/output, run args
- Q: disease: should go elsewhere, maybe with data source
  = [later] disease determines target name
  = consider valiation: some diseases "belong" with some models

## model_config = SimpleNamespace(...)
- belong elsewhere?: sources = ["nhsn"],
-

## both namespaces have model params within in them
- ex: bagging setup: for GBM
- ex: num_samples: flusion
- ex: move from run config to model config:
        num_warmup = 2000,
        num_samples = 2000,
        num_chains = 1

## "--short_run",
- every model should have one (excluding sarix?)


## "disease"
- currently no target param in run config
- but: disease is used for target naming. it's also passed to load_data()


# idmodels
- _format_as_flusight_output(): similar functionality in sarix
e.g.,
        preds_df["target_end_date"] = preds_df["wk_end_date"] + pd.to_timedelta(7*preds_df["horizon"], unit="days")
        preds_df["reference_date"] = ref_date
        preds_df["horizon"] = (pd.to_timedelta(preds_df["target_end_date"].dt.date - ref_date).dt.days / 7).astype(int)
        preds_df["target"] = target_name
- if "nhsn" in self.model_config.sources:
- Q: should run() eventually become generic? assumes features are first-class objects (i.e., ModelConfig or similar has a list of features to add)?


# new classes
## ModelConfig:
- Contains params related to theoretical construction of the chosen model.
- Ex: num_draws, bagging.
- [needs discussion] Q: how data source is integrated with modeling. vs. loading data.


## RunConfig:
- Contains params related to determining internal model operation and model output format.


## smells?
- run(): using model_config values

**Comment by matthewcornell** [comment]
Rough notes from today's code review w/ @lshandross and @trobacker .

# implementation order
- arguments for both top-down (operational-models & idmodels, vs. iddata)
- Q: nick: plans to merge his model code changes?
  - idmodels (FluSight and flu metrocast repos)


# iddata notes
- Idea: add source classes?
- Q: which transformations, if any, should move to idmodels? as we move forward: ask Nick, Evan
  -  terminology: pre-processing (happens at source) vs. feature engineering (happens at model level)?
      - ex: load_agg_transform_ilinet() vs. inc_trans_scale_factor(), respectively
- NEED!: documentation of iddata output table format across diseases
  - Q: should format varies by disease?
  - Q: ditto for pre-processing?
- "process response variable": Q: should this happen here or in idmodels?
- load_us_census(): only used by modeling-based transformations. put elsewhere?
- "disease": some data sources are disease-specific
  - Q: a DiseaseEnum, etc?
- validation needs to centralize (e.g., valid_diseases)


# DataSource
- diseases: [DS.Flu, DS.Covid19, ...]
- as_of: (supported universally?)
- drop_pandemic_seasons?
- TBC


# consider
- arbitrary sources that are simply data frames
- delphi epidata
- scenario projections as a source
- sources: kwargs smell. ex: flusurvnet_kwargs=None, nhsn_kwargs=None, ilinet_kwargs=None, nssp_kwargs
- features: probably should not include any feature engineering - should go in idmodels
- transformations: similar concerns as features. be careful for nssp data source (currently using default one for nhsn)


# REF
NSSP: https://www.cdc.gov/nssp/php/about/index.html
NHSN: https://www.cdc.gov/nhsn/datastat/index.html

**Comment by nickreich** [comment]
## Feature Taxonomy for ML Models

_From a code review of `idmodels`, `iddata`, and `mchub_gbqr` (with Claude Code, Feb 2026). Posting here to inform the refactoring discussion._

We identified four distinct categories of features/data that flow into the GBQR models. Each has different engineering implications for how easy it is to add new instances, where the code lives, and what abstractions would help.

### Category 1: Target-derived features

**What:** Temporal and spatial summaries of the incidence signal (`inc_trans_cs`).

**Examples:**
- Taylor polynomial coefficients (degree 1-2, windows 3-6 weeks) — captures level, velocity, acceleration of the epidemic curve
- Rolling means (2-week, 4-week trailing)
- Temporal lags (lag-1, lag-2 of all the above)
- Directional wave features (distance-weighted averages of neighboring locations' incidence)

**Where the code lives:** `idmodels/preprocess.py` → calls `timeseriesutils` for windowed Taylor coefficients, rolling means, and lags.

**Adding a new one requires:** Changing window sizes, polynomial degrees, or adding new summary statistics in `preprocess.py`. Low engineering cost — it's mostly config.

### Category 2: Structural metadata

**What:** Properties of the observation unit. Plays **two roles**: (a) used as features directly in the model, and (b) used as parameters that define how Category 1 features are computed.

**As features:** one-hot location, one-hot source, one-hot agg_level, `log_pop`, `season_week`, `delta_xmas`

**As parameters for Category 1:** `location` is the grouping variable for per-location rolling means/lags. Lat/lon coordinates (from `state_centroids.csv`) define the spatial weight matrix for directional wave features.

**Adding a new one requires:** A lookup table or static dataset (e.g., urbanicity, population density). One-time merge, minimal pipeline work. But note: new Category 1 features can create new demands on Category 2 — e.g., the wave features required lat/lon that wasn't previously needed.

### Category 3: External covariates

**What:** Data from outside the surveillance system, merged by location + date, adding **columns** to the feature matrix.

**Examples:**
- Weather (Open-Meteo API): 6 raw features × (1 current + 2 lags + 1 rolling avg) = 24 features, z-score standardized
- Potential future additions: wastewater concentrations, Google Trends, school calendar

**Where the code lives:** `mchub_gbqr/data_loader.py` for weather fetching and merge logic.

**Adding a new one requires:** A data fetch pipeline, merge strategy (by what keys? what temporal granularity?), lag/rolling-window feature expansion, missing data handling, and normalization. **Highest engineering cost per feature.** Weather is the template.

### Category 4: Supplementary incidence streams

**What:** Alternative surveillance signals that provide additional training signal. Unlike Category 3, these add **rows** (not columns) to the training data and then go through the same Category 1 feature engineering pipeline.

**Examples:** ILINet, NHSN, FluSurvNet, NSSP — all loaded via `iddata.DiseaseDataLoader` from S3.

**How they work:** Each source gets location-prefixed (e.g., `ilinet_US`, `nhsn_01`) to avoid collisions, concatenated with the primary MCHub data, then power-transformed → scaled/centered → Taylor polynomials → lags → etc. The one-hot `source` column lets the model learn source-specific patterns.

**Adding a new one requires:** A new loader method in `iddata/loader.py`, S3 data pipeline, prefix scheme, and season filtering. Medium cost, but well-templated by existing sources.

### Summary table

| Category | Adds | Code location | Cost to add new | Examples |
|----------|------|---------------|-----------------|----------|
| 1. Target-derived | Features (columns) | `idmodels/preprocess.py` | Low (config) | Taylor polynomials, lags, wave features |
| 2. Structural metadata | Features + parameters | `idmodels/gbqr.py`, lookup tables | Low (one-time) | location, log_pop, lat/lon |
| 3. External covariates | Features (columns) | `mchub_gbqr/data_loader.py` | High (new pipeline) | Weather |
| 4. Supplementary incidence | Training rows | `iddata/loader.py` | Medium (templated) | ILINet, NHSN, NSSP |

### Implications for refactoring

- Categories 1 and 2 are well-contained in `idmodels` already. The main improvement would be making the feature engineering more configurable (e.g., window sizes and polynomial degrees as config rather than hardcoded in `preprocess.py`).
- Category 3 is where the most friction exists for adding new data. Each new external covariate currently requires custom plumbing in the model-specific repo. A generic "external covariate" interface (fetch → align → lag/roll → normalize → merge) could reduce this.
- Category 4 is reasonably well-abstracted by `iddata` already, though the kwargs patterns for source-specific options could be cleaner (as noted in previous comments on this discussion).
- The distinction between rows (Category 4) vs columns (Category 3) is architecturally important — they have fundamentally different merge strategies and different impacts on the feature matrix shape.

**Reply by matthewcornell** [reply]
This is great. I expect @lshandross and @trobacker can make better sense of it. From my perspective, it would be helpful if we could converge on the same terminology. I esp. refer to today's meeting (see notes above) where the three of us use the terms:

> _pre-processing_ (happens at source) vs. _feature engineering_ (happens at model level)

**Reply by nickreich** [reply]
Some comments, post-refactor:

1. Target-derived features: really belong in specific models.
2. Structural metadata: population data exists now as "ancillary data" in iddata. Some geospatial data (e.g. lat/lon for state centroids) still lives in idmodels in, e.g., src/idmodels/data. **Maybe we should make centroids available centrally from iddata as "ancillary data".**
3. External covariates: the iddata/idmodels refactor hasn't really dealt with this, other than possibly allowing this to be called through an iddata dataloader object, but that doesn't really work b/c the column structure assumed by DiseaseDataLoader may not be appropriate here (e.g., maybe multiple columns with different relevant covariates, not just one incidence column). So for now, the thinking this that we haven't dealt with loading this kind of data explicitly and that it would need to be written using custom code for each model. Perhaps at a later date we could generalize the concepts.
4. Supplementary Incidence streams: this is handled in the refactor, and in https://github.com/reichlab/idmodels/pull/27 which allows for primary/secondary data sources.

**Comment by matthewcornell** [comment]
I re-ran Claude on our new `SimpleNamespace`-free codebase, prompting it to focus on `iddata`, think that might be our next target for refactoring. I've included only the **Proposed Changes** section.

----

## Proposed Changes (in priority order)

### Step 1 — Extract S3 file-selection helper

Pull the repeated glob → sort → filter-by-as_of → pick-last pattern into a single private method `_find_versioned_s3_file(s3_prefix, as_of)`. Eliminates 3x duplication.

### Step 2 — Extract constants module (`constants.py`)

Move pandemic seasons list, default FluSurv-NET locations, S3 bucket name, and the `data_raw` base URL into a `constants.py`. Named constants make the logic readable and centralize changes.

### Step 3 — Introduce a `DataSource` protocol/ABC

Define a minimal `DataSource` protocol in a new `sources/base.py`:

```python
class DataSource(Protocol):
    name: str
    def load(self, **kwargs) -> pd.DataFrame: ...
```

Then move each source's loading logic into its own class in `sources/`:

- `sources/nhsn.py` → `NHSNSource`
- `sources/ilinet.py` → `ILINetSource`
- `sources/flusurvnet.py` → `FluSurvNetSource`
- `sources/nssp.py` → `NSSPSource`

Each class encapsulates its own versioning logic, valid diseases, etc. Adding a new source is now just adding a new file that implements the protocol.

### Step 4 — Replace kwargs explosion with a `source_kwargs` dict

Change:

```python
load_data(flusurvnet_kwargs=..., nhsn_kwargs=..., ilinet_kwargs=..., nssp_kwargs=...)
```

to:

```python
load_data(source_kwargs: dict[str, dict] | None = None)
```

where keys are source names. This makes the signature stable regardless of how many sources exist.

### Step 5 — Extract transformation pipeline

Move the fourth-root transform + 95th-percentile scaling + centering block out of `load_data()` into a standalone `transform.py` with an `apply_power_transform(df, power_transform)` function. The pre-scaling offsets in `load_agg_transform_*` should be named constants or parameters.

### Step 6 — Slim down `DiseaseDataLoader`

After the above, `DiseaseDataLoader` (or a renamed `DataLoader`) becomes a thin orchestrator: it holds the registry of available sources, calls `source.load()` for each requested source, merges results, joins census data, and calls `apply_power_transform()`. Supporting data loaders (`load_us_census`, `load_fips_mappings`, `load_hosp_burden`) can remain as methods or move to a `reference.py` module.

---

## Target module structure

```
src/iddata/
├── __init__.py
├── constants.py          # NEW: pandemic seasons, default locations, S3 constants
├── loader.py             # SLIMMED: orchestration only (~100 lines)
├── transform.py          # NEW: apply_power_transform()
├── utils.py              # UNCHANGED
└── sources/
    ├── __init__.py       # NEW
    ├── base.py           # NEW: DataSource protocol
    ├── nhsn.py           # NEW: NHSNSource (from loader.py)
    ├── ilinet.py         # NEW: ILINetSource (from loader.py)
    ├── flusurvnet.py     # NEW: FluSurvNetSource (from loader.py)
    └── nssp.py           # NEW: NSSPSource (from loader.py)
```

**Comment by lshandross** [comment]
Refactor summary and potential remaining items https://github.com/reichlab/decisions/blob/main/project-posters/iddata-idmodels-refactor/refactor_summary.md
